### PR TITLE
AddressMapper client

### PR DESCRIPTION
**Pull request by eduardonunesp**

### CODE REVIEW

File: src/address_mapper.ts [hunk 1]
```diff
@@ -49,19 +50,22 @@ export class AddressMapper {
...
     mappingIdentityRequest.setTo(to.MarshalPB())
 
-    const { r, s, v } = this._signIdentityMapping(from, to, privateKey)
-    const mode = toBuffer(1) as Buffer // geth
-    mappingIdentityRequest.setSignature(Buffer.concat([mode, r, s, toBuffer(v)]))
+    const hash = soliditySha3(
+      [
+        { type: 'address', value: from.local.toString().slice(2) },
+        { type: 'address', value: to.local.toString().slice(2) }
+      ].join('')
```
Comment (enlight) [inline src/address_mapper.ts]: This `join()` doesn't make sense.

File: src/address_mapper.ts [hunk 2]
```diff
@@ -49,19 +50,22 @@ export class AddressMapper {
     )
   }
 
-  private _signIdentityMapping(from: Address, to: Address, privateKey: Uint8Array) {
-    const hash = keccak256([from.toString(), to.toString()])
-    return ecsign(hash, privateKey)
-  }
-
-  async addIdentityMapping(from: Address, to: Address, privateKey: Uint8Array) {
+  async addIdentityMapping(from: Address, to: Address, web3: Web3, ethAddress: string) {
```
Comment (enlight) [inline src/address_mapper.ts]: A `Web3Signer` instance should be passed in as a parameter, you shouldn't create it in here.

Comment (enlight) [reply src/address_mapper.ts]: Also this function should be named `addIdentityMappingAsync`

File: src/address_mapper.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  CallTx
+} from './proto/loom_pb'
+import { CryptoUtils } from '.'
+
+export class AddressMapper {
+  private _client: Client
+  private _privateKey: Uint8Array
+  private _addressMapperContract?: Contract
+
+  constructor(client: Client, privateKey: Uint8Array) {
```
Comment (enlight) [inline src/address_mapper.ts]: Get rid of this private key too, you only need the caller address.

File: src/address_mapper.ts (hunk 1 above)
Comment (eduardonunesp) [reply src/address_mapper.ts]: @enlight the Sha3 result should be the `from` and `to` concatenated isn't?

File: src/plasma-cash/solidity-helpers.ts [hunk 4]
```diff
@@ -31,10 +31,14 @@ export class Web3Signer {
...
+    const isMetamask = ['01', '00'].indexOf(sig.substring(128, 130)) === -1
+
     const r = ethutil.toBuffer('0x' + sig.substring(0, 64)) as Buffer
     const s = ethutil.toBuffer('0x' + sig.substring(64, 128)) as Buffer
-    const v = ethutil.toBuffer(parseInt(sig.substring(128, 130), 16) + 27) as Buffer
-    const mode = ethutil.toBuffer(1) as Buffer // mode = geth
+    const v = ethutil.toBuffer(
+      parseInt(sig.substring(128, 130), 16) + (isMetamask ? 0 : 27)
+    ) as Buffer
+    const mode = ethutil.toBuffer(isMetamask ? 3 : 1) as Buffer // mode = geth
```
Comment (enlight) [inline src/plasma-cash/solidity-helpers.ts]: `mode == 3` isn't going to work, it needs to be 1.

As for the `isMetamask` bit, let's just do this:
```js
let v = parseInt(sig.substring(128, 130), 16)
if (v === 0 || v ===1) {
  v += 27
}
```

**Comment by eduardonunesp** [comment]
Removed the private key from constructor, passing the address, also added a factory static method to create the contract and return `AddressMapper`  ready to use

**Pull request merged**